File: Lights/Ward/Android/app/src/main/java/com/dbooker/livreopengl/WardMaterial.java
package com.dbooker.livreopengl;

// Définition de la classe WardMaterial
// voir http://www.opengl-tutorial.org/intermediate-tutorials/tutorial-13-normal-mapping/

// superclasses et classes nécessaires

import static android.opengl.GLES20.*;

import com.dbooker.livreopengl.glmatrix.mat3;
import com.dbooker.livreopengl.glmatrix.mat4;
import com.dbooker.livreopengl.glmatrix.vec2;
import com.dbooker.livreopengl.glmatrix.vec3;
import com.dbooker.livreopengl.glmatrix.vec4;
import com.dbooker.livreopengl.material.VBOset;
import com.dbooker.livreopengl.mesh.MeshVertex;
import com.dbooker.livreopengl.material.Material;
import com.dbooker.livreopengl.utils.Utils;


/**
 * Cette fonction définit un matériau pour dessiner le maillage
 */
public class WardMaterial extends Material
{
    // caractéristiques du matériau
    private vec3 m_DiffuseColor;
    private vec3 m_SpecularColor;
    private vec2 m_Alpha;

    // coordonnées et couleurs des sources de lumières
    private vec4[] m_LightsPositions;
    private vec3[] m_LightsColors;

    // emplacement des variables du shader
    int m_LightsPositionsLoc = -1;
    int m_LightsColorsLoc = -1;


    /**
     * Cette classe définit un matériau pour dessiner le maillage
     * @param lights_count : nombre de lampes à gérer
     * @param diffuse : fournir un vec3 définissant la couleur diffuse du matériau
     * @param specular : fournir un vec3 définissant la couleur spéculaire du matériau
     * @param alpha : vec2 coefficient de Ward
     */
    public WardMaterial(int lights_count, vec3 diffuse, vec3 specular, vec2 alpha) throws Exception
    {
        super("WardMaterial");

        // caractéristiques du matériau
        m_DiffuseColor = diffuse;
        m_SpecularColor = specular;
        m_Alpha = alpha;

        // coordonnées et couleurs des sources de lumières
        m_LightsPositions = new vec4[lights_count];
        m_LightsColors = new vec3[lights_count];
        for (int i=0; i<lights_count; i++) {
            m_LightsPositions[i] = vec4.create();
            m_LightsColors[i] = vec3.create();
        }

        // compiler le shader
        compileShader();
    }


    /**
     * définit la position d'une lampe
     * @param num : numéro de la lampe
     * @param position : vec4 indiquant où elle se trouve dans l'espace caméra
     * @param color : vec3 indiquant sa couleur et son intensité
     */
    public void setLightPositionColor(int num, vec4 position, vec3 color)
    {
        vec4.copy(m_LightsPositions[num], position);
        vec3.copy(m_LightsColors[num], color);
    }


    /**
     * retourne le source du Vertex Shader
     */
    @Override
    protected String getVertexShader()
    {
        String srcVertexShader =
            "#version 310 es\n" +
            "in vec3 glVertex;\n" +
            "in vec3 glNormal;\n" +
            "in vec3 glTangent;\n" +
            "out vec4 frgPosition;\n" +
            "out vec3 frgNormal;\n" +
            "out vec3 frgTangent;\n" +
            "uniform mat4 mat4ModelView;\n" +
            "uniform mat4 mat4Projection;\n" +
            "uniform mat3 mat3Normal;\n" +
            "void main()\n" +
            "{\n" +
            "    frgPosition = mat4ModelView * vec4(glVertex, 1.0);\n" +
            "    gl_Position = mat4Projection * frgPosition;\n" +
            "    frgNormal = mat3Normal * glNormal;\n" +
            "    frgTangent = mat3Normal * glTangent;\n" +
            "}";
        return srcVertexShader;
    }


    /**
     * retourne le source du Fragment Shader
     */
    @Override
    protected String getFragmentShader()
    {
        int lights_count = m_LightsPositions.length;

        String srcFragmentShader =
            "#version 310 es\n" +
            "precision mediump float;\n" +
            "\n" +
            "// informations géométriques\n" +
            "in vec4 frgPosition;\n" +
            "in vec3 frgNormal;\n" +
            "in vec3 frgTangent;\n" +
            "\n" +
            "// sources de lumière\n" +
            "uniform vec4 LightsPositions["+lights_count+"];\n" +
            "uniform vec3 LightsColors["+lights_count+"];\n" +
            "\n" +
            "// caractéristiques du matériau\n" +
            "const vec3 Kd = "+vec3.str(m_DiffuseColor)+";\n" +
            "const vec3 Ks = "+vec3.str(m_SpecularColor)+";\n" +
            "const vec2 alpha = "+vec2.str(m_Alpha)+";\n" +
            "\n" +
            "const float pi = 3.14159;\n" +
            "\n" +
            "out vec4 glFragColor;\n" +
            "\n" +
            "void main()\n" +
            "{\n" +
            "    // reconstruire le repère de Frenet normalisé\n" +
            "    vec3 N = normalize(frgNormal);\n" +
            "    vec3 T = normalize(frgTangent);\n" +
            "    vec3 B = cross(N, T);\n" +
            "\n" +
            "    // direction du vecteur Vue au niveau du fragment\n" +
            "    vec3 V = normalize(-frgPosition.xyz);\n" +
            "    float dotNV = clamp(dot(N, V), 0.0, 1.0);\n" +
            "\n" +
            "    // couleur finale = cumul des éclairements\n" +
            "    vec3 sum = vec3(0.0, 0.0, 0.0);\n" +
            "\n" +
            "    // boucle sur les lampes\n" +
            "    for (int i=0; i<"+lights_count+"; i++) {\n" +
            "\n" +
            "        // vecteur allant du point à la lumière\n" +
            "        vec3 L = normalize(LightsPositions[i].xyz - frgPosition.xyz);\n" +
            "\n" +
            "        // modulation de l'éclairement type Lambert\n" +
            "        float dotNL = clamp(dot(N, L), 0.0, 1.0);\n" +
            "\n" +
            "        // demi-vecteur H entre L et V, normalisé (!)\n" +
            "        vec3 H = normalize(L + V);\n" +
            "        float dotNH = clamp(dot(N, H), 0.0, 1.0);\n" +
            "\n" +
            "        // éclairement spéculaire Ward\n" +
            "        float hts = dot(H, T) / alpha.s;\n" +
            "        float hbs = dot(H, B) / alpha.t;\n" +
            "        float beta = -2.0 * (hts*hts + hbs*hbs)/(1.0 + dotNH);\n" +
            "        float s = exp(beta) / (4.0 * pi * alpha.s * alpha.t * sqrt(dotNL*dotNV));\n" +
            "\n" +
            "        // cumul des contributions lumineuses\n" +
            "        sum += LightsColors[i] * ( (Kd + Ks * s) * dotNL );\n" +
            "    }\n" +
            "\n" +
            "    glFragColor = vec4(sum, 1.0);\n" +
            "}";
        return srcFragmentShader;
    }


    /**
     * recompile le shader du matériau
     */
    @Override
    protected void compileShader() throws Exception
    {
        // appeler la méthode de la superclasse
        super.compileShader();

        // déterminer où sont les variables uniform
        m_LightsPositionsLoc = glGetUniformLocation(m_ShaderId, "LightsPositions");
        m_LightsColorsLoc  = glGetUniformLocation(m_ShaderId, "LightsColors");
    }


    /**
     * crée et retourne un VBOset pour ce matériau, afin qu'il soit rempli par un maillage
     * @return le VBOset du matériau
     */
    @Override
    public VBOset createVBOset()
    {
        // créer le VBOset et spécifier les noms des attribute nécessaires à ce matériau
        VBOset vboset = super.createVBOset();
        vboset.addAttribute(MeshVertex.ID_ATTR_NORMAL,   Utils.VEC3, "glNormal");
        vboset.addAttribute(MeshVertex.ID_ATTR_TANGENT,  Utils.VEC3, "glTangent");
        return vboset;
    }


    /**
     * Cette méthode active le matériau : met en place son shader,
     * fournit les variables uniform qu'il demande
     * @param mat4Projection : mat4 contenant la projection
     * @param mat4ModelView : mat4 contenant la transformation vers la caméra
     */
    @Override
    public void enable(mat4 mat4Projection, mat4 mat4ModelView)
    {
        // appeler la méthode de la superclasse
        super.enable(mat4Projection, mat4ModelView);

        // initialisation des lampes
        vec4.glUniform(m_LightsPositionsLoc, m_LightsPositions);
        vec3.glUniform(m_LightsColorsLoc,  m_LightsColors);
    }


    /**
     * Cette méthode supprime les ressources allouées
     */
    @Override
    public void destroy()
    {
        super.destroy();
    }
}

File: Lights/Ward/WebGL/WardMaterial.js
// Définition de la classe WardMaterial
// voir http://www.opengl-tutorial.org/intermediate-tutorials/tutorial-13-normal-mapping/

// superclasses et classes nécessaires
Requires("libs/Material/Material.js");
Requires("libs/Material/VBOset.js");


/**
 * Cette fonction définit un matériau pour dessiner le maillage
 */
class WardMaterial extends Material
{
    /**
     * Cette classe définit un matériau pour dessiner le maillage
     * @param lights_count : nombre de lampes à gérer
     * @param diffuse : fournir un vec3 définissant la couleur diffuse du matériau
     * @param specular : fournir un vec3 définissant la couleur spéculaire du matériau
     * @param alpha : vec2 coefficient de Ward
     */
    constructor(lights_count, diffuse, specular, alpha)
    {
        super("WardMaterial");

        // caractéristiques du matériau
        this.m_DiffuseColor = diffuse;
        this.m_SpecularColor = specular;
        this.m_Alpha = alpha;

        // coordonnées et couleurs des sources de lumières
        this.m_LightsPositions = [];
        this.m_LightsColors = [];
        for (let i=0; i<lights_count; i++) {
            this.m_LightsPositions.push(vec4.create());
            this.m_LightsColors.push(vec3.create());
        }

        // compiler le shader
        this.compileShader();
    }


    /**
     * définit la position d'une lampe
     * @param num : numéro de la lampe
     * @param position : vec4 indiquant où elle se trouve dans l'espace caméra
     * @param color : vec3 indiquant sa couleur et son intensité
     */
    setLightPositionColor(num, position, color)
    {
        vec4.copy(this.m_LightsPositions[num], position);
        vec3.copy(this.m_LightsColors[num], color);
    }


    /**
     * retourne le source du Vertex Shader
     */
    getVertexShader()
    {
        let srcVertexShader = dedent
            `#version 300 es
            in vec3 glVertex;
            in vec3 glNormal;
            in vec3 glTangent;
            out vec4 frgPosition;
            out vec3 frgNormal;
            out vec3 frgTangent;
            uniform mat4 mat4ModelView;
            uniform mat4 mat4Projection;
            uniform mat3 mat3Normal;
            void main()
            {
                frgPosition = mat4ModelView * vec4(glVertex, 1.0);
                gl_Position = mat4Projection * frgPosition;
                frgNormal = mat3Normal * glNormal;
                frgTangent = mat3Normal * glTangent;
            }`;
        return srcVertexShader;
    }


    /**
     * retourne le source du Fragment Shader
     */
    getFragmentShader()
    {
        let lights_count = this.m_LightsPositions.length;

        let srcFragmentShader = dedent
            `#version 300 es
            precision mediump float;

            // informations géométriques
            in vec4 frgPosition;
            in vec3 frgNormal;
            in vec3 frgTangent;

            // sources de lumière
            uniform vec4 LightsPositions[${lights_count}];
            uniform vec3 LightsColors[${lights_count}];

            // caractéristiques du matériau
            const vec3 Kd = ${vec3.str(this.m_DiffuseColor)};
            const vec3 Ks = ${vec3.str(this.m_SpecularColor)};
            const vec2 alpha = ${vec2.str(this.m_Alpha)};

            const float pi = 3.14159;

            out vec4 glFragColor;

            void main()
            {
                // reconstruire le repère de Frenet normalisé
                vec3 N = normalize(frgNormal);
                vec3 T = normalize(frgTangent);
                vec3 B = cross(N, T);

                // direction du vecteur Vue au niveau du fragment
                vec3 V = normalize(-frgPosition.xyz);
                float dotNV = clamp(dot(N, V), 0.0, 1.0);

                // couleur finale = cumul des éclairements
                vec3 sum = vec3(0.0, 0.0, 0.0);

                // boucle sur les lampes
                for (int i=0; i<${lights_count}; i++) {

                    // vecteur allant du point à la lumière
                    vec3 L = normalize(LightsPositions[i].xyz - frgPosition.xyz);

                    // modulation de l'éclairement type Lambert
                    float dotNL = clamp(dot(N, L), 0.0, 1.0);

                    // demi-vecteur H entre L et V, normalisé (!)
                    vec3 H = normalize(L + V);
                    float dotNH = clamp(dot(N, H), 0.0, 1.0);

                    // éclairement spéculaire Ward
                    float hts = dot(H, T) / alpha.s;
                    float hbs = dot(H, B) / alpha.t;
                    float beta = -2.0 * (hts*hts + hbs*hbs)/(1.0 + dotNH);
                    float s = exp(beta) / (4.0 * pi * alpha.s * alpha.t * sqrt(dotNL*dotNV));

                    // cumul des contributions lumineuses
                    sum += LightsColors[i] * ( (Kd + Ks * s) * dotNL );
                }

                glFragColor = vec4(sum, 1.0);
            }`
        return srcFragmentShader;
    }


    /**
     * recompile le shader du matériau
     */
    compileShader()
    {
        // appeler la méthode de la superclasse
        super.compileShader();

        // déterminer où sont les variables uniform
        this.m_LightsPositionsLoc = gl.getUniformLocation(this.m_ShaderId, "LightsPositions");
        this.m_LightsColorsLoc  = gl.getUniformLocation(this.m_ShaderId, "LightsColors");
    }


    /**
     * crée et retourne un VBOset pour ce matériau, afin qu'il soit rempli par un maillage
     * @return le VBOset du matériau
     */
    createVBOset()
    {
        // créer le VBOset et spécifier les noms des attribute nécessaires à ce matériau
        let vboset = super.createVBOset();
        vboset.addAttribute(MeshVertex.ID_ATTR_NORMAL,   Utils.VEC3, "glNormal");
        vboset.addAttribute(MeshVertex.ID_ATTR_TANGENT,  Utils.VEC3, "glTangent");
        return vboset;
    }


    /**
     * Cette méthode active le matériau : met en place son shader,
     * fournit les variables uniform qu'il demande
     * @param mat4Projection : mat4 contenant la projection
     * @param mat4ModelView : mat4 contenant la transformation vers la caméra
     */
    enable(mat4Projection, mat4ModelView)
    {
        // appeler la méthode de la superclasse
        super.enable(mat4Projection, mat4ModelView);

        // initialisation des lampes
        vec4.glUniform(this.m_LightsPositionsLoc, this.m_LightsPositions);
        vec3.glUniform(this.m_LightsColorsLoc,  this.m_LightsColors);
    }


    /**
     * Cette méthode supprime les ressources allouées
     */
    destroy()
    {
        super.destroy();
    }
}
